### PR TITLE
dm: use basedb/baseconn in DM & make sure all db connection error contains correct scope

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

- make most db related code use `basedb` or `baseconn`
- make sure all db connection error contains correct scope, i.e. upstream or downstream
- most code change of this pr is code rename or code move, so it's not that big

Issue Number: ref #4287

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
`None`.
```

### CODE REVIEW

File: dm/config/dbconfig/config.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/BurntSushi/toml"
+	"github.com/pingcap/tiflow/dm/config/security"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"github.com/pingcap/tiflow/dm/pkg/terror"
+	"github.com/pingcap/tiflow/dm/pkg/utils"
+	"go.uber.org/zap"
+)
+
+// AdjustDBTimeZone force adjust session `time_zone`.
+func AdjustDBTimeZone(config *DBConfig, timeZone string) {
```
Comment (D3Hunter) [inline dm/config/dbconfig/config.go:30]: moved from `config` package to avoid import cycle

File: dm/config/subtask.go [hunk 2]
```diff
@@ -58,97 +59,6 @@ const (
 	TiDBLightningCheckpointPrefix = "tidb_lightning_checkpoint_"
 )
 
-var defaultMaxIdleConns = 2
```
Comment (D3Hunter) [inline dm/config/subtask.go:61]: moved into sub package `dbconfig`

File: dm/pkg/conn/db.go [hunk 3]
```diff
@@ -132,48 +135,21 @@ func GetSlaveServerID(ctx context.Context, db *sql.DB) (map[uint32]struct{}, err
...
-			return nil, terror.DBErrorAdapt(err, terror.ErrDBDriverError)
+			return nil, terror.DBErrorAdapt(err, db.Scope, terror.ErrDBDriverError)
 		}
 		serverIDs[uint32(serverIDUInt)] = struct{}{}
 	}
 	return serverIDs, nil
 }
 
-// GetGlobalVariable gets server's global variable.
-func GetGlobalVariable(ctx context.Context, db *sql.DB, variable string) (value string, err error) {
```
Comment (D3Hunter) [inline dm/pkg/conn/db.go:149]: this code duplicate with code in `conn/utils.go`, only paramater is different, so it's deleted, below `getVariable` too

File: dm/pkg/utils/common.go [hunk 4]
```diff
@@ -62,155 +56,6 @@ func TrimQuoteMark(s string) string {
 	return s
 }
 
-// FetchAllDoTables returns all need to do tables after filtered (fetches from upstream MySQL).
```
Comment (D3Hunter) [inline dm/pkg/utils/common.go:65]: moved into `pkg/conn` to use `basedb`, below code too

File: dm/pkg/conn/basedb.go [hunk 5]
```diff
@@ -232,17 +278,27 @@ func (d *BaseDB) DoTxWithRetry(tctx *tcontext.Context, queries []string, args []
...
 	defer d.mu.Unlock()
 	delete(d.conns, conn)
 	return conn.close()
 }
 
-// CloseBaseConnWithoutErr close the base connect and output a warn log if meets an error.
-func CloseBaseConnWithoutErr(d *BaseDB, conn *BaseConn) {
-	if err1 := d.CloseBaseConn(conn); err1 != nil {
+// ForceCloseConn release BaseConn resource from BaseDB, and close BaseConn completely(not return to the connection pool).
+func (d *BaseDB) ForceCloseConn(conn *BaseConn) error {
```
Comment (D3Hunter) [inline dm/pkg/conn/basedb.go:291]: rename it to `force-close` meet its real meaning

File: dm/pkg/conn/basedb.go [hunk 6]
```diff
@@ -232,17 +278,27 @@ func (d *BaseDB) DoTxWithRetry(tctx *tcontext.Context, queries []string, args []
 	return err
 }
 
-// CloseBaseConn release BaseConn resource from BaseDB, and close BaseConn.
-func (d *BaseDB) CloseBaseConn(conn *BaseConn) error {
+// CloseConn release BaseConn resource from BaseDB, and returns the connection to the connection pool,
+// has the same meaning of sql.Conn.Close
+func (d *BaseDB) CloseConn(conn *BaseConn) error {
```
Comment (D3Hunter) [inline dm/pkg/conn/basedb.go:283]: this `close` has the same meaning as sql.Conn.Close, i.e. put it into connection pool.

File: dm/pkg/conn/baseconn.go [hunk 7]
```diff
@@ -259,11 +275,20 @@ func (conn *BaseConn) ApplyRetryStrategy(tctx *tcontext.Context, params retry.Pa
...
 func (conn *BaseConn) close() error {
 	if conn == nil || conn.DBConn == nil {
 		return nil
 	}
+	return conn.DBConn.Close()
+}
+
+// forceClose will close the underlying connection completely,
+// should not be used by functions other than BaseDB.ForceCloseConn.
+func (conn *BaseConn) forceClose() error {
```
Comment (D3Hunter) [inline dm/pkg/conn/baseconn.go:288]: rename to meet its real meaning

File: dm/pkg/conn/baseconn.go [hunk 8]
```diff
@@ -259,11 +275,20 @@ func (conn *BaseConn) ApplyRetryStrategy(tctx *tcontext.Context, params retry.Pa
 	return conn.RetryStrategy.Apply(tctx, params, operateFn)
 }
 
-// close should not be used by functions other than BaseDB.CloseBaseConn.
+// close returns the connection to the connection pool, has the same meaning of sql.Conn.Close
```
Comment (D3Hunter) [inline dm/pkg/conn/baseconn.go:278]: this one has the same meaning of `sql.Conn.Close()`, i.e. put into connection pool

**Review by lance6716 — APPROVED** [review]

File: dm/pkg/conn/baseconn.go [hunk 9]
```diff
@@ -72,17 +72,33 @@ import (
...
-	return &BaseConn{conn, strategy}
+	return &BaseConn{
+		DBConn:        conn,
+		Scope:         scope,
+		RetryStrategy: strategy,
+	}
+}
+
+// NewBaseConnForTest builds BaseConn to connect real DB for test.
+func NewBaseConnForTest(conn *sql.Conn, strategy retry.Strategy) *BaseConn {
```
Comment (okJiang) [inline dm/pkg/conn/baseconn.go:93]: ditto

File: dm/pkg/conn/basedb_test.go [hunk 10]
```diff
@@ -75,18 +74,18 @@ func TestFailDBPing(t *testing.T) {
 	require.NoError(t, err)
 	defer l.Close()
 
-	cfg := &config.DBConfig{User: "root", Host: "127.0.0.1", Port: port}
+	cfg := &dbconfig.DBConfig{User: "root", Host: "127.0.0.1", Port: port}
 	cfg.Adjust()
 	impl := &DefaultDBProviderImpl{}
-	db, err := impl.Apply(cfg)
+	db, err := impl.Apply(UpstreamDBConfig(cfg))
```
Comment (okJiang) [inline dm/pkg/conn/basedb_test.go:80]: can use `GetUpstreamDB`

File: dm/pkg/conn/basedb.go [hunk 11]
```diff
@@ -232,17 +278,35 @@ func (d *BaseDB) DoTxWithRetry(tctx *tcontext.Context, queries []string, args []
...
+func (d *BaseDB) ForceCloseConn(conn *BaseConn) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	delete(d.conns, conn)
+	return conn.forceClose()
+}
+
+// ForceCloseConnWithoutErr close the connection completely(not return to the conn pool),
+// and output a warning log if meets an error.
+func ForceCloseConnWithoutErr(d *BaseDB, conn *BaseConn) {
```
Comment (okJiang) [inline dm/pkg/conn/basedb.go:308]: seems `func (d *BaseDB) ForceCloseConnWithoutErr)` is better, but it doesn't matter.

File: dm/pkg/checker/conn_checker.go [hunk 12]
```diff
@@ -56,7 +56,7 @@ func (c *connNumberChecker) check(ctx context.Context, checkerName string, neede
 		return result
 	}
 	//nolint:errcheck
-	defer c.toCheckDB.CloseBaseConn(baseConn)
+	defer c.toCheckDB.ForceCloseConn(baseConn)
```
Comment (okJiang) [inline dm/pkg/checker/conn_checker.go:59]: `ForceCloseConnWithoutErr` is better

File: dm/pkg/v1dbschema/schema.go [hunk 13]
```diff
@@ -47,7 +47,7 @@ func UpdateSchema(tctx *tcontext.Context, db *conn.BaseDB, cfg *config.SubTaskCo
 		return terror.ErrFailUpdateV1DBSchema.Delegate(err)
 	}
 	defer func() {
-		err2 := db.CloseBaseConn(dbConn)
+		err2 := db.ForceCloseConn(dbConn)
```
Comment (okJiang) [inline dm/pkg/v1dbschema/schema.go:50]: `ForceCloseConnWithoutErr` is better

File: dm/chaos/cases/db.go [hunk 14]
```diff
@@ -48,7 +48,7 @@ func createDBConn(ctx context.Context, db *conn.BaseDB, currDB string) (*dbConn,
 		baseConn: c,
 		currDB:   currDB,
 		resetFunc: func(ctx context.Context, baseConn *conn.BaseConn) (*conn.BaseConn, error) {
-			err2 := db.CloseBaseConn(baseConn)
+			err2 := db.ForceCloseConn(baseConn)
```
Comment (okJiang) [inline dm/chaos/cases/db.go:51]: `ForceCloseConnWithoutErr` is better

File: dm/loader/db.go [hunk 15]
```diff
@@ -215,7 +223,7 @@ func createConns(tctx *tcontext.Context, cfg *config.SubTaskConfig,
 			return nil, nil, terror.WithScope(err, terror.ScopeDownstream)
 		}
 		resetBaseConnFn := func(tctx *tcontext.Context, baseConn *conn.BaseConn) (*conn.BaseConn, error) {
-			err := baseDB.CloseBaseConn(baseConn)
+			err := baseDB.ForceCloseConn(baseConn)
```
Comment (okJiang) [inline dm/loader/db.go:226]: `ForceCloseConnWithoutErr` is better

**Review by okJiang — APPROVED** [review]

File: dm/pkg/conn/basedb.go [hunk 16]
```diff
@@ -146,14 +173,33 @@ type BaseDB struct {
...
+		DB:            db,
+		conns:         conns,
+		Retry:         &retry.FiniteRetryStrategy{},
+		Scope:         scope,
+		doFuncInClose: doFuncInClose,
+	}
+}
+
+// NewBaseDBForTest returns *BaseDB object for test.
+func NewBaseDBForTest(db *sql.DB, doFuncInClose ...func()) *BaseDB {
```
[the remark this answers is not in the document]
Comment (D3Hunter) [reply dm/pkg/conn/basedb.go:194]: put into test file cannot be referenced cross package

**Comment by D3Hunter** [comment]
wait until 6.5 code freeze

**Comment by D3Hunter** [comment]
/run-integration-test tidb=f5487e3f2596cdd63ae9840606c7fff3fb539e03
/run-kafka-integration-test tidb=f5487e3f2596cdd63ae9840606c7fff3fb539e03

**Pull request merged**